File: src/IO/hdf5MeshWriter.cc
#ifndef HDF5MESHWRITER_CC
#define HDF5MESHWRITER_CC

#include "hdf5MeshWriter.hh"

template <int dim>
HDF5MeshWriter<dim>::HDF5MeshWriter(const std::string& baseName, const NodeList* nodeList, const std::vector<std::string>& fieldNames)
    : baseName(baseName), nodeList(nodeList), fieldNames(fieldNames) {}

template <int dim>
void 
HDF5MeshWriter<dim>::write(const std::string& fileName) {
    std::string fullFileName = baseName + fileName;
    H5::H5File file(fullFileName, H5F_ACC_TRUNC);

    writePointData(file);

    file.close();
}

template <int dim>
void 
HDF5MeshWriter<dim>::writePointData(H5::Group& group) {
    auto positions = nodeList->position<dim>();
    if (!positions) {
        std::cerr << "Position field is not available in the NodeList!" << std::endl;
        return;
    }

    hsize_t dims[2] = {nodeList->size(), dim};
    H5::DataSpace dataspace(2, dims);

    H5::DataSet dataset = group.createDataSet("Position", H5::PredType::NATIVE_DOUBLE, dataspace);

    std::vector<double> data(nodeList->size() * dim);
    for (unsigned int idx = 0; idx < nodeList->size(); ++idx) {
        Lin::Vector<dim> pos = positions->getValue(idx);
        for (int i = 0; i < dim; ++i) {
            data[idx * dim + i] = pos[i];
        }
    }

    dataset.write(data.data(), H5::PredType::NATIVE_DOUBLE);
}

#endif

File: src/IO/hdf5MeshWriter.py
from PYB11Generator import *

@PYB11template("dim")
class HDF5MeshWriter:
    def pyinit(self,
               baseName="std::string&",
               nodeList="NodeList*",
               fieldNames="std::vector<std::string>&"):
        return
    def write(self,fileName="std::string&"):
        return

HDF5MeshWriter1d = PYB11TemplateClass(HDF5MeshWriter,
                              template_parameters = ("1"),
                              cppname = "HDF5MeshWriter<1>",
                              pyname = "HDF5MeshWriter1d",
                              docext = " (1D).")
HDF5MeshWriter2d = PYB11TemplateClass(HDF5MeshWriter,
                              template_parameters = ("2"),
                              cppname = "HDF5MeshWriter<2>",
                              pyname = "HDF5MeshWriter2d",
                              docext = " (2D).")
HDF5MeshWriter3d = PYB11TemplateClass(HDF5MeshWriter,
                              template_parameters = ("3"),
                              cppname = "HDF5MeshWriter<3>",
                              pyname = "HDF5MeshWriter3d",
                              docext = " (3D).") 